File: ios/SpentyAI/SpentyAI/Features/Customers/CustomerRepository.swift
import Foundation

// MARK: - Models

struct CustomerInvoice: Codable, Identifiable {
    let id: String
    let invoiceNumber: String?
    let date: Date?
    let dueDate: Date?
    let total: Double?
    let amountPaid: Double?
    let status: String?

    enum CodingKeys: String, CodingKey {
        case id = "invoiceId"
        case invoiceNumber
        case date
        case dueDate
        case total
        case amountPaid
        case status
    }
}

struct CustomerInvoicesResponse: Codable {
    let items: [CustomerInvoice]
    let total: Int?
    let customerName: String?
}

struct CustomerPayload: Codable {
    let name: String
    let email: String?
    let phone: String?
    let gstin: String?
    let billingAddress: String?
    let shippingAddress: String?
}

struct CustomerListResponse: Codable {
    let items: [Customer]
    let total: Int?
}

// MARK: - Repository

struct CustomerRepository {

    private let api = APIClient.shared

    // MARK: - CRUD

    func fetchAll() async throws -> [Customer] {
        let response: CustomerListResponse = try await api.get(APIEndpoints.customers)
        return response.items
    }

    func create(_ payload: CustomerPayload) async throws -> Customer {
        try await api.post(APIEndpoints.customers, body: payload)
    }

    func update(id: String, _ payload: CustomerPayload) async throws -> Customer {
        try await api.put(APIEndpoints.customer(id), body: payload)
    }

    func delete(id: String) async throws {
        let _: MessageResponse = try await api.delete(APIEndpoints.customer(id))
    }

    // MARK: - Invoices

    func fetchInvoices(customerId: String) async throws -> [CustomerInvoice] {
        let response: CustomerInvoicesResponse = try await api.get(APIEndpoints.customerInvoices(customerId))
        return response.items
    }
}

File: android-native/app/src/main/java/com/spentyai/app/features/customers/CustomerRepository.kt
package com.spentyai.app.features.customers

import com.spentyai.app.core.models.Customer
import com.spentyai.app.core.models.Invoice
import com.spentyai.app.core.network.ApiClient
import com.spentyai.app.core.network.ApiResult
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

class CustomerRepository(private val apiClient: ApiClient) {

    // CRUD

    suspend fun fetchAll(): ApiResult<List<Customer>> =
        apiClient.safeApiCall { apiClient.endpoints.getCustomers() }

    suspend fun fetch(id: String): ApiResult<Customer> =
        apiClient.safeApiCall { apiClient.endpoints.getCustomer(id) }

    suspend fun create(payload: JsonObject): ApiResult<Customer> =
        apiClient.safeApiCall { apiClient.endpoints.createCustomer(payload) }

    suspend fun update(id: String, payload: JsonObject): ApiResult<Customer> =
        apiClient.safeApiCall { apiClient.endpoints.updateCustomer(id, payload) }

    suspend fun delete(id: String): ApiResult<JsonObject> =
        apiClient.safeApiCall { apiClient.endpoints.deleteCustomer(id) }

    // Invoices for a customer
    suspend fun fetchCustomerInvoices(customerId: String): ApiResult<List<Invoice>> =
        apiClient.safeApiCall { apiClient.endpoints.getInvoices(customerId = customerId) }
}

fun buildCustomerPayload(
    name: String,
    email: String?,
    phone: String?,
    gstin: String?,
    billingAddress: String?,
    shippingAddress: String?
): JsonObject = buildJsonObject {
    put("name", name)
    email?.let { put("email", it) }
    phone?.let { put("phone", it) }
    gstin?.let { put("tax_id", it) }
    billingAddress?.let {
        put("billing_address", buildJsonObject {
            put("street", it)
        })
    }
    shippingAddress?.let {
        put("shipping_address", buildJsonObject {
            put("street", it)
        })
    }
}
